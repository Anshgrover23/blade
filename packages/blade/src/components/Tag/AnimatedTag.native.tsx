import React from 'react';
import { Tag } from './Tag';
import type { AnimatedTagProps } from './types';
import BaseBox from '~components/Box/BaseBox';

const AnimatedTag = ({
  children,
  currentTagIndex,
  activeTagIndex,
  onDismiss,
}: AnimatedTagProps): React.ReactElement => {
  return (
    <BaseBox>
      <Tag
        _isVirtuallyFocussed={currentTagIndex === activeTagIndex}
        _isTagInsideInput={true}
        onDismiss={() => {
          onDismiss({ tagIndex: currentTagIndex, tagName: children });
        }}
        marginRight="spacing.3"
        marginY="spacing.2"
      >
        {children}
      </Tag>
    </BaseBox>
  );
};

export { AnimatedTag };
