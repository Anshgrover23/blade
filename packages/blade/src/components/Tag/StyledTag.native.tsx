import type { BaseBoxProps } from '~components/Box/BaseBox';
import BaseBox from '~components/Box/BaseBox';

const StyledTag = (props: BaseBoxProps): React.ReactElement => {
  return (
    <BaseBox display="flex" flexDirection="row" flexWrap="nowrap" alignSelf="center" {...props} />
  );
};

export { StyledTag };
