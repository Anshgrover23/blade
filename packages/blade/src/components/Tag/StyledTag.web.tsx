import type { BaseBoxProps } from '~components/Box/BaseBox';
import BaseBox from '~components/Box/BaseBox';

const StyledTag = (props: BaseBoxProps): React.ReactElement => {
  return <BaseBox display="inline-block" {...props} />;
};

export { StyledTag };
