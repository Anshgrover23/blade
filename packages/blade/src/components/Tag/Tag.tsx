import React from 'react';
import { StyledTag } from './StyledTag';
import { Box } from '~components/Box';
import type { StyledPropsBlade } from '~components/Box/styledProps';
import { getStyledProps } from '~components/Box/styledProps';
import { IconButton } from '~components/Button/IconButton';
import type { IconComponent } from '~components/Icons';
import { CloseIcon } from '~components/Icons';
import { Text } from '~components/Typography';
import type { StringChildrenType, TestID } from '~src/_helpers/types';
import { metaAttribute, MetaConstants } from '~utils';

type TagProps = {
  /**
   * Decides the size of Tag
   *
   * @default medium
   */
  size?: 'medium' | 'large';

  /**
   * Leading icon for your Tag
   */
  icon?: IconComponent;

  /**
   * Callback when close icon on Tag is clicked
   */
  onDismiss?: ({ value }: { value: StringChildrenType }) => void;

  /**
   * Text that renders inside Tag
   */
  children: StringChildrenType;

  /**
   * Disable tag
   */
  isDisabled?: boolean;
} & StyledPropsBlade &
  TestID;

const Tag = ({
  size = 'medium',
  icon: Icon,
  onDismiss,
  children,
  isDisabled,
  testID,
  ...styledProps
}: TagProps): React.ReactElement | null => {
  const [isVisible, setIsVisible] = React.useState(true);

  if (!isVisible) {
    return null;
  }

  const textColor = isDisabled
    ? 'surface.text.placeholder.lowContrast'
    : 'surface.text.subtle.lowContrast';

  return (
    <StyledTag
      backgroundColor="brand.gray.a100.lowContrast"
      borderRadius="max"
      padding={
        size === 'medium'
          ? ['spacing.1', 'spacing.2', 'spacing.1', 'spacing.3']
          : ['spacing.2', 'spacing.3', 'spacing.2', 'spacing.4']
      }
      {...getStyledProps(styledProps)}
      {...metaAttribute({ name: MetaConstants.Tag, testID })}
    >
      <Box display="flex" flexDirection="row" flexWrap="nowrap" alignItems="center">
        {Icon ? <Icon color={textColor} size="small" marginRight="spacing.2" /> : null}
        <Text marginRight="spacing.2" color={textColor} size="small">
          {children}
        </Text>
        <IconButton
          size="small"
          icon={CloseIcon}
          accessibilityLabel={`Close ${children} tag`}
          isDisabled={isDisabled}
          onClick={() => {
            setIsVisible(false);
            onDismiss?.({ value: children });
          }}
        />
      </Box>
    </StyledTag>
  );
};

export { Tag, TagProps };
