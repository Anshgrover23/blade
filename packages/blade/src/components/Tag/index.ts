export * from './Tag';
export { TagProps } from './types';
