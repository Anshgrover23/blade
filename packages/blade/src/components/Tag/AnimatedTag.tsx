import React from 'react';
import styled, { css, FlattenSimpleInterpolation, keyframes } from 'styled-components';
import BaseBox from '~components/Box/BaseBox';
import { makeMotionTime, useTheme } from '~utils';
import { Tag } from './Tag';
import { AnimatedTagProps } from './types';

const tagDissappearKeyframe = keyframes`
  100% {
    opacity: 1;
    max-width: 140px;
  }

  100% {
    opacity: 0;
    max-width: 0px;
    max-height: 0px;
  }
`;

const tagShowKeyframe = keyframes`
  0% {
    opacity: 0;
  }

  100% {
    opacity: 1;
  }
`;

const AnimatedTagContainer = styled(BaseBox)<{
  transition: FlattenSimpleInterpolation;
}>(
  (props) => css`
    ${props.transition};
    display: inline-block;
    max-width: 140px;
  `,
);

const AnimatedTag = ({
  children,
  currentTagIndex,
  activeTagIndex,
  onDismiss,
  tagsLength,
}: AnimatedTagProps) => {
  const [isTagVisible, setIsTagVisible] = React.useState(true);
  const { theme } = useTheme();
  const prevSelectionsLength = React.useRef<number>();

  const hideTagTransition = css`
    animation: ${tagDissappearKeyframe} ${makeMotionTime(theme.motion.duration.xquick)}
      ${String(theme.motion.easing.exit.effective)};
  `;

  const showTagTransition = css`
    animation: ${tagShowKeyframe} ${makeMotionTime(theme.motion.duration.xquick)}
      ${String(theme.motion.easing.entrance.effective)};
  `;

  const noTransition = css`
    animation: none;
  `;

  const isTagRemoved = prevSelectionsLength.current
    ? prevSelectionsLength.current > tagsLength
    : false;

  return (
    <AnimatedTagContainer
      style={{ opacity: isTagVisible ? 1 : 0 }}
      onAnimationEnd={() => {
        if (!isTagVisible) {
          setIsTagVisible(true);
          onDismiss({ tagIndex: currentTagIndex, tagName: children });
        }
      }}
      transition={
        isTagRemoved ? noTransition : isTagVisible ? showTagTransition : hideTagTransition
      }
    >
      <Tag
        _isVirtuallyFocussed={currentTagIndex === activeTagIndex}
        _isTagInsideInput={true}
        onDismiss={() => {
          prevSelectionsLength.current = tagsLength;
          setIsTagVisible(false);
        }}
        marginRight="spacing.3"
        marginY="spacing.2"
      >
        {children}
      </Tag>
    </AnimatedTagContainer>
  );
};

export { AnimatedTag };
