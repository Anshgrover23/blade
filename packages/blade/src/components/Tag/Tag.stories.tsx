import React from 'react';
import type { DOMAttributes } from 'react';
import type { Meta } from '@storybook/react';
import { Title } from '@storybook/addon-docs';
import type { TagProps } from './Tag';
import { Tag } from './Tag';
import { Sandbox } from '~src/_helpers/storybook/Sandbox';
import StoryPageWrapper from '~src/_helpers/storybook/StoryPageWrapper';
import { getStyledPropsArgTypes } from '~components/Box/BaseBox/storybookArgTypes';
import { Box } from '~components/Box';
import { Button } from '~components/Button';
import { TextInput } from '~components/Input/TextInput';
import { isReactNative } from '~utils';
import { PlusIcon } from '~components/Icons';

const Page = (): React.ReactElement => {
  return (
    <StoryPageWrapper
      figmaURL={{
        paymentTheme:
          'https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?node-id=28435%3A581488',
        bankingTheme:
          'https://www.figma.com/file/sAdplk2uYnI2ILnDKUxycW/Blade---Banking-Dark?node-id=15234%3A480939',
      }}
      componentName="Tags"
      componentDescription="These are set of interactive keywords that help organise & categorise objects. Tags can be added or removed from an object by the users."
    >
      <Title>Usage</Title>
      <Sandbox>
        {`
        import { Tag } from '@razorpay/blade/components';
        
        function App(): JSX.Element {
          return (
            <Tag onDismiss={({ value }) => console.log('Tag dismissed', value)}>
              Unpaid
            </Tag>
          )
        }

        export default App;
        `}
      </Sandbox>
    </StoryPageWrapper>
  );
};

export default {
  title: 'Components/Tag',
  component: Tag,
  argTypes: {
    ...getStyledPropsArgTypes(),
  },
  parameters: {
    docs: {
      page: Page,
    },
  },
} as Meta<TagProps>;

export const Default = (props: TagProps): React.ReactElement => <Tag {...props} />;

Default.args = {
  children: 'Unpaid',
  onDismiss: ({ value }) => console.log('dismiss tag', value),
} as TagProps;

const CrossPlatformForm = ({
  children,
  onSubmit,
}: {
  children: React.ReactElement[];
  onSubmit: DOMAttributes<HTMLFormElement>['onSubmit'];
}): React.ReactElement => {
  if (isReactNative()) {
    return <Box>{children}</Box>;
  }

  return <form onSubmit={onSubmit}>{children}</form>;
};

export const ControlledTags = (props: TagProps): React.ReactElement => {
  const [inputValue, setInputValue] = React.useState('');
  const [tags, setTags] = React.useState<string[]>([]);

  const addTag = (): void => {
    // Add input value to tags and clear the input value
    setTags([...tags, inputValue]);
    setInputValue('');
  };

  const removeTag = (tagName: TagProps['children']): void => {
    setTags(tags.filter((tagNameValue) => tagNameValue !== tagName));
  };

  return (
    <Box>
      <Box paddingY="spacing.4">
        {tags.map((tagName) => (
          <Tag
            key={tagName}
            {...props}
            marginLeft="spacing.2"
            onDismiss={({ value }) => removeTag(value)}
          >
            {tagName}
          </Tag>
        ))}
      </Box>
      <Box>
        <CrossPlatformForm
          onSubmit={(e) => {
            e.preventDefault();
            addTag();
          }}
        >
          <TextInput
            label="Tag Label"
            value={inputValue}
            onChange={({ value }) => setInputValue(value ?? '')}
            {...{ onSubmit: isReactNative() ? () => addTag() : undefined }}
          />
          <Button
            icon={PlusIcon}
            iconPosition="right"
            variant="secondary"
            marginTop="spacing.2"
            type="submit"
          >
            Create Tag
          </Button>
        </CrossPlatformForm>
      </Box>
    </Box>
  );
};
