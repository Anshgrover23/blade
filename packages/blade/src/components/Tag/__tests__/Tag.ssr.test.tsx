/* eslint-disable @typescript-eslint/explicit-function-return-type */
/* eslint-disable @typescript-eslint/no-unsafe-return */
import React from 'react';
import { Tag } from '../Tag';
import renderWithSSR from '~src/_helpers/testing/renderWithSSR.web';

describe('<Tag />', () => {
  it('should render tag', () => {
    const { container } = renderWithSSR(<Tag>in:User</Tag>);
    expect(container).toMatchSnapshot();
  });
});
