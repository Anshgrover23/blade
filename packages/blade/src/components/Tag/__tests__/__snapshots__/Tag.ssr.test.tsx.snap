// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> should render tag 1`] = `"<div id=\\"root\\" data-reactroot=\\"\\"><div data-blade-component=\\"tag\\" class=\\"BaseBox-bmPWx Tag__FocussableTag-jmwb4r-0 iZePvJ\\"><div data-blade-component=\\"box\\" class=\\"BaseBox-bmPWx eEphev\\"><div data-blade-component=\\"box\\" class=\\"BaseBox-bmPWx dnZxUO\\"><div data-blade-component=\\"box\\" class=\\"BaseBox-bmPWx hbrzAK\\"><p class=\\"StyledBaseText-dVBfTO kDKzNc\\" data-blade-component=\\"text\\">in:User</p></div><button type=\\"button\\" aria-label=\\"Close in:User tag\\" data-blade-component=\\"icon-button\\" class=\\"StyledIconButtonweb__StyledButton-sc-1f4cg7d-0 gvuEeM\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"12px\\" viewBox=\\"0 0 24 24\\" width=\\"12px\\" fill=\\"none\\" class=\\"Svgweb__StyledSvg-vcmjs8-0\\"><path d=\\"M18.7071 6.70711C19.0976 6.31658 19.0976 5.68342 18.7071 5.29289C18.3166 4.90237 17.6834 4.90237 17.2929 5.29289L12 10.5858L6.70711 5.29289C6.31658 4.90237 5.68342 4.90237 5.29289 5.29289C4.90237 5.68342 4.90237 6.31658 5.29289 6.70711L10.5858 12L5.29289 17.2929C4.90237 17.6834 4.90237 18.3166 5.29289 18.7071C5.68342 19.0976 6.31658 19.0976 6.70711 18.7071L12 13.4142L17.2929 18.7071C17.6834 19.0976 18.3166 19.0976 18.7071 18.7071C19.0976 18.3166 19.0976 17.6834 18.7071 17.2929L13.4142 12L18.7071 6.70711Z\\" fill=\\"currentColor\\" data-blade-component=\\"svg-path\\"></path></svg></button></div><div data-blade-component=\\"box\\" class=\\"BaseBox-bmPWx cfbWPC\\"><div data-blade-component=\\"box\\" class=\\"BaseBox-bmPWx hbrzAK\\"><p class=\\"StyledBaseText-dVBfTO kDKzNc\\" data-blade-component=\\"text\\">in:User</p></div><button type=\\"button\\" aria-label=\\"Close in:User tag\\" data-blade-component=\\"icon-button\\" class=\\"StyledIconButtonweb__StyledButton-sc-1f4cg7d-0 gvuEeM\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"12px\\" viewBox=\\"0 0 24 24\\" width=\\"12px\\" fill=\\"none\\" class=\\"Svgweb__StyledSvg-vcmjs8-0\\"><path d=\\"M18.7071 6.70711C19.0976 6.31658 19.0976 5.68342 18.7071 5.29289C18.3166 4.90237 17.6834 4.90237 17.2929 5.29289L12 10.5858L6.70711 5.29289C6.31658 4.90237 5.68342 4.90237 5.29289 5.29289C4.90237 5.68342 4.90237 6.31658 5.29289 6.70711L10.5858 12L5.29289 17.2929C4.90237 17.6834 4.90237 18.3166 5.29289 18.7071C5.68342 19.0976 6.31658 19.0976 6.70711 18.7071L12 13.4142L17.2929 18.7071C17.6834 19.0976 18.3166 19.0976 18.7071 18.7071C19.0976 18.3166 19.0976 17.6834 18.7071 17.2929L13.4142 12L18.7071 6.70711Z\\" fill=\\"currentColor\\" data-blade-component=\\"svg-path\\"></path></svg></button></div></div></div></div>"`;

exports[`<Tag /> should render tag 2`] = `
<div
  data-reactroot=""
  id="root"
>
  <div
    class="BaseBox-bmPWx Tag__FocussableTag-jmwb4r-0 iZePvJ"
    data-blade-component="tag"
  >
    <div
      class="BaseBox-bmPWx eEphev"
      data-blade-component="box"
    >
      <div
        class="BaseBox-bmPWx dnZxUO"
        data-blade-component="box"
      >
        <div
          class="BaseBox-bmPWx hbrzAK"
          data-blade-component="box"
        >
          <p
            class="StyledBaseText-dVBfTO kDKzNc"
            data-blade-component="text"
          >
            in:User
          </p>
        </div>
        <button
          aria-label="Close in:User tag"
          class="StyledIconButtonweb__StyledButton-sc-1f4cg7d-0 gvuEeM"
          data-blade-component="icon-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="Svgweb__StyledSvg-vcmjs8-0"
            data-blade-component="icon"
            fill="none"
            height="12px"
            viewBox="0 0 24 24"
            width="12px"
          >
            <path
              d="M18.7071 6.70711C19.0976 6.31658 19.0976 5.68342 18.7071 5.29289C18.3166 4.90237 17.6834 4.90237 17.2929 5.29289L12 10.5858L6.70711 5.29289C6.31658 4.90237 5.68342 4.90237 5.29289 5.29289C4.90237 5.68342 4.90237 6.31658 5.29289 6.70711L10.5858 12L5.29289 17.2929C4.90237 17.6834 4.90237 18.3166 5.29289 18.7071C5.68342 19.0976 6.31658 19.0976 6.70711 18.7071L12 13.4142L17.2929 18.7071C17.6834 19.0976 18.3166 19.0976 18.7071 18.7071C19.0976 18.3166 19.0976 17.6834 18.7071 17.2929L13.4142 12L18.7071 6.70711Z"
              data-blade-component="svg-path"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
      <div
        class="BaseBox-bmPWx cfbWPC"
        data-blade-component="box"
      >
        <div
          class="BaseBox-bmPWx hbrzAK"
          data-blade-component="box"
        >
          <p
            class="StyledBaseText-dVBfTO kDKzNc"
            data-blade-component="text"
          >
            in:User
          </p>
        </div>
        <button
          aria-label="Close in:User tag"
          class="StyledIconButtonweb__StyledButton-sc-1f4cg7d-0 gvuEeM"
          data-blade-component="icon-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="Svgweb__StyledSvg-vcmjs8-0"
            data-blade-component="icon"
            fill="none"
            height="12px"
            viewBox="0 0 24 24"
            width="12px"
          >
            <path
              d="M18.7071 6.70711C19.0976 6.31658 19.0976 5.68342 18.7071 5.29289C18.3166 4.90237 17.6834 4.90237 17.2929 5.29289L12 10.5858L6.70711 5.29289C6.31658 4.90237 5.68342 4.90237 5.29289 5.29289C4.90237 5.68342 4.90237 6.31658 5.29289 6.70711L10.5858 12L5.29289 17.2929C4.90237 17.6834 4.90237 18.3166 5.29289 18.7071C5.68342 19.0976 6.31658 19.0976 6.70711 18.7071L12 13.4142L17.2929 18.7071C17.6834 19.0976 18.3166 19.0976 18.7071 18.7071C19.0976 18.3166 19.0976 17.6834 18.7071 17.2929L13.4142 12L18.7071 6.70711Z"
              data-blade-component="svg-path"
              fill="currentColor"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
